### ISSUE TITLE
Accept empty arrays (ACCEPT_EMPTY_ARRAY_AS_NULL_OBJECT)

### ISSUE THREAD

**Issue by james-woods** [ngs-doo/dsl-json#13]
The following JSON breaks the parser:

```
{
	"@timestamp": "2016-09-05T12:44:46.000Z",
	"url": {
		"analyzed": "/-/-/call-1111",
		"path": "/-/-/call-1111",
		"query": []
	},
	"event": "view",
	"origin": "detail",
	"session": ["-9035111172716044880", "6141211111163737100", "-6891111112685400843", "-275111111849250600"],
	"@version": "1"
}
```
(Query is a nested JSON object, but when it's empty it is written out as an empty array.)
It would be nice to have an option like ACCEPT_EMPTY_ARRAY_AS_NULL_OBJECT when deserializing JSON.

**Comment by zapov** [comment]
If I'm understanding you correctly you want parser to cope with invalid JSON?

It should be 

    "query": {} 

in the JSON instead of array?

I planned on supporting custom converters per property... I guess thats something which could be handled more easily after I add that feature.

**Comment by james-woods** [comment]
Yes, in an ideal world it should look like that. The problem is that these are historic logs from an ElasticSearch cluster - when there is nothing, query is an empty array, but when there is one element it looks exactly like
```
"query": {"bla": 1,...}
```

Putting the empty array version in an JSON validator works, parsing the whole thing via Jackson and with the option `ACCEPT_EMPTY_ARRAY_AS_NULL_OBJECT` works too, but Jackson is slowish..

**Comment by zapov** [comment]
Currently you can use a fake type to simulate such behavior. Something like:

    @CompiledJson
    public static class Elastic {
        @JsonAttribute(name = "@timestamp")
        public OffsetDateTime timestamp;
        public Url url;
        public String event;
    }

    public static class Url {
        public String analyzed;
        public ElasticQuery query;
    }

    @JsonConverter(target = ElasticQuery.class)
    public static abstract class ElasticQueryConverter {
        public static JsonReader.ReadObject<BrokenQuery> BROKEN_READER;
        public static JsonReader.ReadObject<ElasticQuery> JSON_READER = new JsonReader.ReadObject<ElasticQuery>() {
            public ElasticQuery read(JsonReader jsonReader) throws IOException {
                if (jsonReader.last() == '[') {
                    jsonReader.getNextToken();
                    return null;
                } else if (jsonReader.last() == '{') {
                    return BROKEN_READER.read(jsonReader);
                }
                return null;
            }
        };
        public static JsonWriter.WriteObject<ElasticQuery> JSON_WRITER = null;
    }

    @CompiledJson
    public static class BrokenQuery extends ElasticQuery {
    }

    public static class ElasticQuery {
        public int bla;
    }

    static class DslJsonWrapper extends DslJson<Object> {
        public <T> JsonReader.ReadObject<T> findReader(Class<T> manifest) {
            return (JsonReader.ReadObject<T>)tryFindReader(manifest);
        }
    }

    public static void main(String[] args) throws IOException {
        DslJsonWrapper dslJson = new DslJsonWrapper();
        ElasticQueryConverter.BROKEN_READER = dslJson.findReader(BrokenQuery.class);
        String json = "{\n" +
                "\t\"@timestamp\": \"2016-09-05T12:44:46.000Z\",\n" +
                "\t\"url\": {\n" +
                "\t\t\"analyzed\": \"/-/-/call-1111\",\n" +
                "\t\t\"path\": \"/-/-/call-1111\",\n" +
                "\t\t\"query\": []\n" +
                "\t},\n" +
                "\t\"event\": \"view\",\n" +
                "\t\"origin\": \"detail\",\n" +
                "\t\"session\": [\"-9035111172716044880\", \"6141211111163737100\", \"-6891111112685400843\", \"-275111111849250600\"],\n" +
                "\t\"@version\": \"1\"\n" +
                "}";
        Elastic e = dslJson.deserialize(Elastic.class, json.getBytes(), json.length());
        System.out.println(e);
    }

I had to wrap DslJson since tryFindReader is not public.
I used a fake type to force Dsl compiler to create appropriate converter and subtype to specify my own converter.
In there you can handle you custom logic.

In the future I plan on allowing custom converters per property... so this should remove the fake type requirement from the story.

**Comment by james-woods** [comment]
Thanks! This workaround fixed this particular issue

**Comment by james-woods** [comment]
The remaining bug is a sporadic Nullpointer exception, when I wrap the whole thing into Google Dataflow. I have the feeling this is related to static classes and likely an concurrency issue. Is dsl-json thread safe used in this example?

**Comment by zapov** [comment]
Hm... NPE on which line? BROKEN_READER?
DslJson is thread safe... but Readers and Writers are not. If you want best perf you need to reuse buffers and Writers in thread local

**Comment by james-woods** [comment]
```
 public JsonLogLine defensiveBuilder(String line, String filename) {
            try {
                byte [] buffer = line.getBytes("UTF-8");

                DslJsonWrapper dslJson = new DslJsonWrapper();
                QueryConverter.BROKEN_READER = dslJson.findReader(BrokenQuery.class);

                log.info(line);
                Elastic elastic = dslJson.deserialize(Elastic.class, buffer, buffer.length);

                String t = elastic.timestamp;
                return new JsonLogLine(
                        filename,
                        TimeStructure.JsonFormat(t),
                        elastic
                );

            }
            catch (IOException e) {
                log.error("IOException parsing entry: " + e.getMessage());
                return INVALID;
            }
            catch(NullPointerException e) {
                log.error("NullPointerException parsing entry: " + e.getMessage());
                return INVALID;
            }
        }

```

This method is called by the framework once per line of input. Each line is a valid JSON object. It usually breaks on the second line of input so I guess my reuse of the Readers (I do not use Writers) is wrong. The NPE is happening directly afterwards when the `JsonLogLine` object, that the method yields is passed to the next method in the framework. I have different branches with different JSON parsers so I'm assuming my code of the parser is causing it, since it's the only difference between the branches.

**Comment by zapov** [comment]
While idealy you should have only one instance of DslJson and assign custom reader only once.. this code shouldnt have concurrency  problems.
